lib/nrf_inbuilt_key/nrf_inbuilt_key.c

```diff
@@ -0,0 +1,307 @@
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	VALIDATE_LENGTH(scratch_buf_len, sizeof(scratch_buf));
+
+	written = snprintf(scratch_buf, scratch_buf_len, "%s,%u,%u\r\n",
+			   NRF_INBUILT_KEY_OP_RD, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+
+	if (written < 0) {
+		return -ENOBUFS;
+	}
+
+	VALIDATE_LENGTH(written, scratch_buf_len);
+	VALIDATE_LENGTH(written, *p_buffer_len);
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, scratch_buf,
+					       sizeof(scratch_buf), &state);
+
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		return translate_error(result);
+	}
+
+	if (result == 0) {
+		struct at_param_list cmng_list;
+
+		at_params_list_init(&cmng_list, 5);
+		at_parser_params_from_str(scratch_buf, NULL, &cmng_list);
+
+		size_t len;
+
+		at_params_size_get(&cmng_list, 4, &len);
+		if (len < *p_buffer) {
+			at_params_list_free(&cmng_list);
+			return -ENOBUFS;
+		}
+		at_params_string_get(&cmng_list, 4, p_buffer, &len);
+		*p_buffer_len = len;
+		at_params_list_free(&cmng_list);
+	}
+
+	return result;
+}
+
+int nrf_inbuilt_key_delete(nrf_sec_tag_t sec_tag,
+			   enum nrf_key_mgnt_cred_type_t cred_type)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	u16_t scratch_buf_len = sizeof(NRF_INBUILT_KEY_OP_RM) +
+				(NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 * 2) +
+				(NRF_INBUILT_KEY_COMMA_SIZE * 2) +
+				NRF_INBUILT_KEY_CR_CHAR_SIZE +
+				NRF_INBUILT_KEY_LF_CHAR_SIZE +
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	VALIDATE_LENGTH(scratch_buf_len, sizeof(scratch_buf));
+
+	written = snprintf(scratch_buf, scratch_buf_len, "%s,%u,%u\r\n",
+			   NRF_INBUILT_KEY_OP_RM, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+
+	VALIDATE_LENGTH(written, scratch_buf_len);
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, NULL, 0, &state);
+
+	if ((result != 0) && (state == AT_CMD_ERROR_CME)) {
+		return translate_error(result);
+	}
+
+	return result;
+}
+
+int nrf_inbuilt_key_permission_set(nrf_sec_tag_t sec_tag,
+				   enum nrf_key_mgnt_cred_type_t cred_type,
+				   u8_t perm_flags)
+{
+	int retval = -EOPNOTSUPP;
+	return retval;
+}
+
+int nrf_inbuilt_key_exists(nrf_sec_tag_t sec_tag,
+			   enum nrf_key_mgnt_cred_type_t cred_type,
+			   bool *p_exists, u8_t *p_perm_flags)
+{
+	int result;
+	int written;
+	enum at_cmd_state state;
+
+	if ((p_exists == NULL) || (p_perm_flags == NULL)) {
+		return -EINVAL;
+	}
+
+	u16_t scratch_buf_len = sizeof(NRF_INBUILT_KEY_OP_LS) +
+				(NRF_INBUILT_KEY_CHAR_SIZE_MAX_INT32 * 2) +
+				(NRF_INBUILT_KEY_COMMA_SIZE * 2) +
+				NRF_INBUILT_KEY_CR_CHAR_SIZE +
+				NRF_INBUILT_KEY_LF_CHAR_SIZE +
+				NRF_INBUILT_KEY_ZERO_TERMINATOR_SIZE;
+
+	VALIDATE_LENGTH(scratch_buf_len, sizeof(scratch_buf));
+
+	written = snprintf(scratch_buf, scratch_buf_len, "%s,%u,%u\r\n",
+			   NRF_INBUILT_KEY_OP_LS, (u32_t)sec_tag,
+			   (u8_t)cred_type);
+
+	VALIDATE_LENGTH(written, scratch_buf_len);
+
+	result = write_at_cmd_with_cme_enabled(scratch_buf, scratch_buf,
+					       sizeof(scratch_buf), &state);
```

Comment:
It's generally scetchy to use the same buffer for input and output. We need to either make the API explicitly document that this is OK. Or use a different buffer for the response.

Reply:
We cannot afford yet another 4K static buffer in response path, which is the max size of any AT command that can provision/read a credential. We will only potentially use the 4k buffer at extreme max in one direction at the time (write or read). We should also trust that the at_cmd driver is working serialized, sending the request, then waiting for the response.

Reply:
This trust should be explicitly documented in the API of the at_cmd driver.

Reply:
I agree, this feature should be explicitly documented in the API, especially that we keep relying on this in various places to save RAM. **I will prepare a PR** with a documentation update today.